extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -222,32 +223,68 @@ Most of the window operations like SUM(), RANK(), RANGE() etc. can be made with
         final Frame frame = inputChannel.read();
         convertRowFrameToRowsAndColumns(frame);
         return ReturnOrAwait.runAgain();
-      } else if (inputChannel.isFinished()) {
-        runAllOpsOnMultipleRac(frameRowsAndCols);
+      }
+
+      if (inputChannel.isFinished()) {
+        // If there are rows pending flush, flush them and run again.
+        if (frameHasRowsPendingFlush()) {
+          return flushRACsAndRunAgain();
+        }
+
+        // If no rows are flushed yet, process all rows.
+        if (rowId.get() == 0) {
+          runAllOpsOnMultipleRac(frameRowsAndCols);
+        }
+
+        // If there are still rows pending after operations, run again.
+        if (frameHasRowsPendingFlush()) {
+          return ReturnOrAwait.runAgain();
+        }
```

Comment:
Responded here: https://github.com/apache/druid/pull/17209#discussion_r1782640897

Reply:
ok - I see

If we get more than twice the work we could handle from the inputChannel we will be overwhelmed by that after some time

instead of all this; it would be better to wrap `runIncrementally` around to flush out stuff first if the writer is not ready to accept more data - and only when its ready let it pick up more.

Reply:
> instead of all this; it would be better to wrap runIncrementally around to flush out stuff first if the writer is not ready to accept more data - and only when its ready let it pick up more.

Can you elaborate more on this?
"wrap runIncrementally" - how do you picture this? 🤔

Reply:
something like:
```
  public ReturnOrAwait<Object> runIncrementally(IntSet readableInputs)
  {
         if (frameHasRowsPendingFlush()) {
          return flushRACsAndRunAgain();
        } else {
           return realRunIncrementally(readableInputs);
        }
  }
```

Reply:
@kgyrtkirk I'd have to spend some time into this, it's difficult to guess what might cause issues with this approach.

But before that, just to be on the same page, can I assume that the current way of deciding whether we need a re-run via `frameHasRowsPendingFlush()` is okay? (This is in regard to https://github.com/apache/druid/pull/17209#discussion_r1782736678 - sorry for the overlapping discussion threads)

```java
   private boolean frameHasRowsPendingFlush()
   {
     return frameWriter != null && frameWriter.getNumRows() > 0;
   }
```

Reply:
I've just used what's already there - yes; if that's when there are data to be written then yes

how the current approach would handle the case when:
* on every call to `runIncrementally` there are `10K` rows are available
* in the meantime its only able to output 1K rows with the framewriter in every round

Reply:
@kgyrtkirk Have updated `runIncrementally()` to look like the following:
```java
   if (frameHasRowsPendingFlush()) {
      return flushRACsAndRunAgain();
    }
   // all other code of runIncrementally()
```